pkg/services/alerting/ticker_test.go

```diff
@@ -1,121 +1,53 @@
-// 	case tick := <-ticker.C:
-// 		t.Fatalf("timer should have stayed at %s, instead it advanced to %s", desiredLast, tick)
-// 	case <-time.NewTimer(wait).C:
-// 		return
-// 	}
-// }
-// }
-//
-// func TestTickerRetro1Hour(t *testing.T) {
-// offset := time.Duration(10) * time.Second
-// last := time.Unix(0, 0)
-// mock := clock.NewMock()
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast := mock.Now().Add(-offset)
-// ticker := NewTicker(last, offset, mock)
-//
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(10)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// }
-//
-// func TestAdvanceWithUpdateOffset(t *testing.T) {
-// offset := time.Duration(10) * time.Second
-// last := time.Unix(0, 0)
-// mock := clock.NewMock()
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast := mock.Now().Add(-offset)
-// ticker := NewTicker(last, offset, mock)
-//
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(10)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// // lowering offset should see a few more ticks
-// offset = time.Duration(5) * time.Second
-// ticker.updateOffset(offset)
-// desiredLast = mock.Now().Add(-offset)
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(9)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// // advancing clock should see even more ticks
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast = mock.Now().Add(-offset)
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(8)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// }
-//
-// func getCase(lastSeconds, offsetSeconds int) (time.Time, time.Duration) {
-// last := time.Unix(int64(lastSeconds), 0)
-// offset := time.Duration(offsetSeconds) * time.Second
-// return last, offset
-// }
-//
-// func TestTickerNoAdvance(t *testing.T) {
-//
-// // it's 00:01:00 now. what are some cases where we don't want the ticker to advance?
-// mock := clock.NewMock()
-// mock.Add(time.Duration(60) * time.Second)
-//
-// type Case struct {
-// 	last   int
-// 	offset int
-// }
-//
-// // note that some cases add up to now, others go into the future
-// cases := []Case{
-// 	{50, 10},
-// 	{50, 30},
-// 	{59, 1},
-// 	{59, 10},
-// 	{59, 30},
-// 	{60, 1},
-// 	{60, 10},
-// 	{60, 30},
-// 	{90, 1},
-// 	{90, 10},
-// 	{90, 30},
-// }
-// for _, c := range cases {
-// 	last, offset := getCase(c.last, c.offset)
-// 	ticker := NewTicker(last, offset, mock)
-// 	assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-// }
-// }
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTicker(t *testing.T) {
+	t.Run("should not drop ticks", func(t *testing.T) {
```

Comment:
Looks good in practice but I don't think this is testing the accumulation of ticks.

What you want to test here is assuming you have 4 ticks that happen at t1, t2, t3, t4 and an interval of N:

1. If t2 takes a duration of t2+N, then t3 is not executed until t2 is done and subsequently t4 is executed at t4+N*2.
2. In the case where t3 is stuck forever (e.g. 5 ticks) it won't advance, piling up _all_ ticks that will eventually execute, in order.

Case #1 is asserting against a single tick misbehaving, whereas case #2 is about how subsequent ticks behave. You can do all of them in one big test but I think it'll be more readable if you split it into two. Make sure you also assert timestamps (that the value of C is what you would expect).

You can probably do #1 by having a loop that will sleep for interval * 2 on iteration or any other approach that you think can work.